### PR TITLE
Run device mesh tests with native funcol enabled

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119104
* #119103
* #119102
* #118498
* #118986
* __->__ #118437
* #118911
* #118910

### Summary

Run the relevant tests in `test/distributed/_tensor/test_dtensor_compile.py` and `test/distributed/test_device_mesh.py` with native funcol enabled, in addition to with them being disabled.

All tests excepts `test_tp_compile_comm_reordering` pass. This is expected because the native funcols have slightly different IRs, so the reordering pass needs to be adjusted. This test is disabled for now.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: test/distributed/test_device_mesh.py [hunk 1]
```diff
@@ -28,6 +28,7 @@
 from torch.testing._internal.common_utils import run_tests
 from torch.testing._internal.distributed._tensor.common_dtensor import (
     DTensorTestBase,
+    NativeFunColMixin,
```
Comment (wanchaol) [inline test/distributed/test_device_mesh.py:31]: `NativeFunColMixin` is this defined somewhere? I couldn't find the definition

File: test/distributed/test_device_mesh.py [hunk 2]
```diff
@@ -726,5 +727,31 @@ def test_all_to_all_nd(self):
                 self.assertEqual(output_tensor, expected_tensor)
 
 
+class DeviceMeshTestWithNativeFunCol(DeviceMeshTest):
+    def setUp(self) -> None:
+        self._prev_native_funcol_enabled = funcol.native_funcol_enabled()
+        funcol.enable_native_funcol()
```
Comment (wanchaol) [inline test/distributed/test_device_mesh.py:733]: I'm wondering if making these two classes change become a single decorator or context manager would make things simpler?

**Review by wanchaol — COMMENTED** [review]
sgtm, but looks like the CI failures are relevant?

**Review by LucasLLC — APPROVED** [review]
lgtm!

**Pull request closed without merging**